File: csharp/Lib/1277.CountSquareSubmatriceswithAllOnes.cs
/*
1277. Count Square Submatrices with All Ones
*/

using System;
using System.Linq;

public class CountSquareSubmatriceswithAllOnes {
    public int CountSquares(int[][] matrix) {
		if (matrix == null || matrix.Length == 0)
			return 0;
		

		return _CountSquare(matrix);
    }

	private int _CountSquare(int[][] matrix) {
		var dp = _GenTable(matrix.Length, matrix[0].Length);

		var count = 0;
		for (int i = 0; i < dp.Length; ++i)
			for (int j = 0; j < dp[i].Length; ++j) {
				if (i == 0 || j == 0) dp[i][j] = matrix[i][j];
				else if (matrix[i][j] == 0) dp[i][j] = 0;
				else dp[i][j] = 1 + new int[] {
					dp[i-1][j],
					dp[i-1][j-1],
					dp[i][j-1],
				}.Min();
				count += dp[i][j];
			}

		return count;
	}

	private int[][] _GenTable(int numRows, int numCols) {
		var table = new int[numRows][];

		for (int i = 0; i < numRows; ++i) {
			table[i] = new int[numCols];
			for (int j = 0; j < numCols; ++j)
				table[i][j] = 0;
		}

		return table;
	}
}

File: cpp/lib/1277.CountSquareSubmatricesWithAllOnes.cpp
// 1277. Count Square Submatrices with All Ones

#include "1277.CountSquareSubmatricesWithAllOnes.h"
#include <algorithm>

vector<vector<int>>* LeetCode1277CountSquareSubmatricesWithAllOnes::GenTable_(int rowSize, int colSize) {
    auto table = new vector<vector<int>>(rowSize, vector<int>(colSize));

    return table;
}

int LeetCode1277CountSquareSubmatricesWithAllOnes::CountSquares_(vector<vector<int>>& matrix) {
    auto rowSize = matrix.size();
    auto colSize = matrix[0].size();
    auto& table = *GenTable_(rowSize, colSize);

    auto count = 0;
    for (int row = 0; row < matrix.size(); ++row) {
        for (int col = 0; col < matrix[row].size(); ++col) {
            if (row == 0 || col == 0) table[row][col] = matrix[row][col];
            else if (matrix[row][col] == 0) table[row][col] = 0;
            else table[row][col] = 1 + std::min({table[row-1][col], table[row-1][col-1], table[row][col-1]});
            count += table[row][col];
        }
    }

    delete &table;
    return count;
}

int LeetCode1277CountSquareSubmatricesWithAllOnes::countSquares(vector<vector<int>>& matrix) {
    if (matrix.size() == 0) return 0;
    return CountSquares_(matrix);
}
